### PR TITLE
Add GLM-4 and Later GLM Model (Draft)

### PR DESCRIPTION
**Pull request by zRzRzRzRzRzRzR**
This is a draft and we will continue work
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

## Who can review?

@ArthurZucker

### CODE REVIEW

**Comment by xianbaoqian** [comment]
Hi @zRzRzRzRzRzRzR ! Thanks for drafting the PR. The workflow has been failing due to the usage of TikToken.  Once the converter scripts converts tiktoken configuration to HF tokenizer configuration, you won't need to import tiktoken during inference at tokenization_glm.py

**Comment by ArthurZucker** [comment]
YEP! Will review today 🤗

**Comment by zRzRzRzRzRzRzR** [comment]
> Hi @zRzRzRzRzRzRzR ! Thanks for drafting the PR. The workflow has been failing due to the usage of TikToken. Once the converter scripts converts tiktoken configuration to HF tokenizer configuration, you won't need to import tiktoken during inference at tokenization_glm.py

Fix this issue now~ Tks

File: docs/source/en/model_doc/glm.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+# GLM
+
+## Overview
+
+The GLM Model was proposed
+in [ChatGLM: A Family of Large Language Models from GLM-130B to GLM-4 All Tools](https://arxiv.org/html/2406.12793v1)
+by GLM Team, THUDM & ZhipuAI. GLM models released with 5 versions, Which are GLM-130B,ChatGLM-6B,ChatGLM2-6B,ChatGLM3-6B
+and GLM-4.
+
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/glm.md:25]: let's add the abstract of the paper here!

File: docs/source/en/model_doc/glm.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+Tips:
+
+- This model was contributed by [THUDM](https://huggingface.co/THUDM). The most recent code can be
+  found [here](https://github.com/thudm/GLM-4).
+
+## GLMConfig
+
+[[autodoc]] GLMConfig
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/glm.md:33]:
```suggestion
[[autodoc]] GlmConfig
```
let's use camel casing everywhere we can!

File: src/transformers/convert_slow_tokenizer.py [hunk 3]
```diff
@@ -855,7 +899,15 @@ def vocab(self, proto):
...
+        vocab += [("ar_AR", 0.0), ("cs_CZ", 0.0), ("de_DE", 0.0), ("en_XX", 0.0), ("es_XX", 0.0), ("et_EE", 0.0),
+                  ("fi_FI", 0.0), ("fr_XX", 0.0), ("gu_IN", 0.0), ("hi_IN", 0.0), ("it_IT", 0.0), ("ja_XX", 0.0),
+                  ("kk_KZ", 0.0), ("ko_KR", 0.0), ("lt_LT", 0.0), ("lv_LV", 0.0), ("my_MM", 0.0), ("ne_NP", 0.0),
+                  ("nl_XX", 0.0), ("ro_RO", 0.0), ("ru_RU", 0.0), ("si_LK", 0.0), ("tr_TR", 0.0), ("vi_VN", 0.0),
+                  ("zh_CN", 0.0), ("af_ZA", 0.0), ("az_AZ", 0.0), ("bn_IN", 0.0), ("fa_IR", 0.0), ("he_IL", 0.0),
+                  ("hr_HR", 0.0), ("id_ID", 0.0), ("ka_GE", 0.0), ("km_KH", 0.0), ("mk_MK", 0.0), ("ml_IN", 0.0),
+                  ("mn_MN", 0.0), ("mr_IN", 0.0), ("pl_PL", 0.0), ("ps_AF", 0.0), ("pt_XX", 0.0), ("sv_SE", 0.0),
+                  ("sw_KE", 0.0), ("ta_IN", 0.0), ("te_IN", 0.0), ("th_TH", 0.0), ("tl_XX", 0.0), ("uk_UA", 0.0),
+                  ("ur_PK", 0.0), ("xh_ZA", 0.0), ("gl_ES", 0.0), ("sl_SI", 0.0)]  # fmt: skip
```
Comment (ArthurZucker) [inline src/transformers/convert_slow_tokenizer.py:910]: this seems unrelated let's revert it

File: src/transformers/convert_slow_tokenizer.py [hunk 4]
```diff
@@ -1263,7 +1315,8 @@ def vocab(self, proto):
             ("<unk>", 0.0),
         ]
         vocab += [(piece.piece, piece.score) for piece in proto.pieces[3:]]
-        vocab += [("<madeupword0>", 0.0), ("<madeupword1>", 0.0), ("<madeupword2>", 0.0), ("<madeupword3>", 0.0), ("<madeupword4>", 0.0), ("<madeupword5>", 0.0), ("<madeupword6>", 0.0)]  # fmt: skip
+        vocab += [("<madeupword0>", 0.0), ("<madeupword1>", 0.0), ("<madeupword2>", 0.0), ("<madeupword3>", 0.0),
+                  ("<madeupword4>", 0.0), ("<madeupword5>", 0.0), ("<madeupword6>", 0.0)]  # fmt: skip
```
Comment (ArthurZucker) [inline src/transformers/convert_slow_tokenizer.py:1319]: same here

File: src/transformers/models/glm/tokenization_glm_fast.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+VOCAB_FILES_NAMES = {
+    "vocab_file": "vocab.json",
+    "merges_file": "merges.txt",
+    "tokenizer_file": "tokenizer.json",
+}
+
+
+class GLMTokenizerFast(PreTrainedTokenizerFast):
+    """
+    Construct a "fast" Qwen2 tokenizer (backed by HuggingFace's *tokenizers* library). Based on byte-level
```
Comment (ArthurZucker) [inline src/transformers/models/glm/tokenization_glm_fast.py:36]: I it a Qwen2 or a GLM tokenizer? 
Also do we really need a new class? IMO just the PreTrainedTokenizerFast should suffice, unless you want the add_eos and add_bos functionnalities

File: src/transformers/models/glm/tokenization_glm_fast.py
Comment (ArthurZucker) [inline src/transformers/models/glm/tokenization_glm_fast.py:1]: to remove let's just used `PreTrainedTokenizerFast`

File: src/transformers/models/glm/tokenization_glm.py
Comment (ArthurZucker) [inline src/transformers/models/glm/tokenization_glm.py:1]: This is basically GPT2Tokenizer, we don't need a new class if the only thing we change is the chat template! Let's use GPT2Tokenizer and GPT2TokenizerFast no?

File: src/transformers/models/glm/modeling_glm.py [hunk 6]
```diff
@@ -0,0 +1,1549 @@
...
+# Copied from transformers.models.llama.modeling_llama._get_unpad_data
+def _get_unpad_data(attention_mask):
+    seqlens_in_batch = attention_mask.sum(dim=-1, dtype=torch.int32)
+    indices = torch.nonzero(attention_mask.flatten(), as_tuple=False).flatten()
+    max_seqlen_in_batch = seqlens_in_batch.max().item()
+    cu_seqlens = F.pad(torch.cumsum(seqlens_in_batch, dim=0, dtype=torch.int32), (1, 0))
+    return (
+        indices,
+        cu_seqlens,
+        max_seqlen_in_batch,
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:74]: we refactored this see #31446

File: src/transformers/models/glm/modeling_glm.py [hunk 7]
```diff
@@ -0,0 +1,1549 @@
...
+    # Split.
+    tensor_list = torch.split(tensor, last_dim_size, dim=last_dim)
+    # Note: torch.split does not create contiguous tensors by default.
+    if contiguous_split_chunks:
+        return tuple(chunk.contiguous() for chunk in tensor_list)
+
+    return tensor_list
+
+
+class SelfAttention(torch.nn.Module):
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:163]:
```suggestion
class GlmSelfAttention(torch.nn.Module):
```

File: src/transformers/models/glm/modeling_glm.py [hunk 8]
```diff
@@ -0,0 +1,1549 @@
...
+        self.qkv_hidden_size = 3 * self.projection_size
+        if self.multi_query_attention:
+            self.num_multi_query_groups_per_partition = config.multi_query_group_num
+            self.qkv_hidden_size = (
+                    self.projection_size + 2 * self.hidden_size_per_attention_head * config.multi_query_group_num
+            )
+        self.query_key_value = nn.Linear(config.hidden_size, self.qkv_hidden_size,
+                                         bias=config.add_bias_linear or config.add_qkv_bias,
+                                         device=device, **_config_to_kwargs(config)
+                                         )
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:191]: the code should be agnostic to this parameter and support both without an if else. THis is done if phi3 for example

File: src/transformers/models/glm/modeling_glm.py [hunk 9]
```diff
@@ -0,0 +1,1549 @@
...
+            )
+            value_layer = value_layer.view(
+                value_layer.size()[:-1]
+                + (self.num_multi_query_groups_per_partition, self.hidden_size_per_attention_head)
+            )
+        else:
+            new_tensor_shape = mixed_x_layer.size()[:-1] + \
+                               (self.num_attention_heads_per_partition,
+                                3 * self.hidden_size_per_attention_head)
+            mixed_x_layer = mixed_x_layer.view(*new_tensor_shape)
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:257]: same comment here, we already have https://github.com/huggingface/transformers/blob/0fdea8607d7e01eb0e38a1ebeb7feee30a22f0cf/src/transformers/models/phi3/modeling_phi3.py#L289-L291 
 
and https://github.com/huggingface/transformers/blob/0fdea8607d7e01eb0e38a1ebeb7feee30a22f0cf/src/transformers/models/phi3/modeling_phi3.py#L324-L333

File: src/transformers/models/glm/modeling_glm.py [hunk 10]
```diff
@@ -0,0 +1,1549 @@
...
+            key_layer = key_layer.contiguous().view(
+                key_layer.size()[:1] + (self.num_attention_heads_per_partition,) + key_layer.size()[3:]
+            )
+            value_layer = value_layer.unsqueeze(2)
+            value_layer = value_layer.expand(
+                -1, -1, self.num_attention_heads_per_partition // self.num_multi_query_groups_per_partition, -1, -1
+            )
+            value_layer = value_layer.contiguous().view(
+                value_layer.size()[:1] + (self.num_attention_heads_per_partition,) + value_layer.size()[3:]
+            )
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:288]: let's just use the https://github.com/huggingface/transformers/blob/0fdea8607d7e01eb0e38a1ebeb7feee30a22f0cf/src/transformers/models/phi3/modeling_phi3.py#L245-L254 repeate kv !

File: src/transformers/models/glm/modeling_glm.py [hunk 11]
```diff
@@ -0,0 +1,1549 @@
...
+
+        # ==================================
+        # core attention computation
+        # ==================================
+
+        context_layer = self.core_attention(query_layer, key_layer, value_layer, attention_mask)
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:294]: why do we separate this? Let's do like we usually do and just implement it here!

File: src/transformers/models/glm/modeling_glm.py [hunk 12]
```diff
@@ -0,0 +1,1549 @@
...
+
+        # Calculate the product of position index and $\theta_i$
+        idx_theta = torch.outer(seq_idx, theta).float()
+
+        cache = torch.stack([torch.cos(idx_theta), torch.sin(idx_theta)], dim=-1).to(dtype=dtype)
+        return cache
+
+    def forward(self, max_seq_len, offset=0):
+        return self.forward_impl(
+            max_seq_len, self.dim, dtype=self.inv_freq.dtype, device=self.inv_freq.device
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:130]: This is not copied from gemma (the code is not per say copied from) 
It says it is an enhence version, what is the difference? It looks mathematically equivalent so let's re-use the proof tested rope from Gemma (and actually copy from here!)_

File: src/transformers/models/glm/modeling_glm.py [hunk 13]
```diff
@@ -0,0 +1,1549 @@
...
+_CHECKPOINT_FOR_DOC = "THUDM/glm-4-9b-chat"
+_CONFIG_FOR_DOC = "GLMConfig"
+
+
+def _config_to_kwargs(args):
+    common_kwargs = {
+        "dtype": args.torch_dtype,
+    }
+    return common_kwargs
+
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:63]: we shoukld not need that

File: src/transformers/models/glm/modeling_glm.py [hunk 14]
```diff
@@ -0,0 +1,1549 @@
...
+            num_attention_heads = self.num_multi_query_groups_per_partition
+        else:
+            num_attention_heads = self.num_attention_heads_per_partition
+        return torch.empty(
+            inference_max_sequence_len,
+            batch_size,
+            num_attention_heads,
+            self.hidden_size_per_attention_head,
+            dtype=dtype,
+            device=device,
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:211]: let's remove this as well if not used

File: src/transformers/models/glm/modeling_glm.py [hunk 15]
```diff
@@ -0,0 +1,1549 @@
...
+            key_layer = key_layer.unsqueeze(2)
+            key_layer = key_layer.expand(
+                -1, -1, self.num_attention_heads_per_partition // self.num_multi_query_groups_per_partition, -1, -1
+            )
+            key_layer = key_layer.contiguous().view(
+                key_layer.size()[:1] + (self.num_attention_heads_per_partition,) + key_layer.size()[3:]
+            )
+            value_layer = value_layer.unsqueeze(2)
+            value_layer = value_layer.expand(
+                -1, -1, self.num_attention_heads_per_partition // self.num_multi_query_groups_per_partition, -1, -1
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:284]: we don't support partitionning as in TP or MP in transformers, so this is more like `num_kv_heads` let's re-use what is common

File: src/transformers/models/glm/modeling_glm.py [hunk 16]
```diff
@@ -0,0 +1,1549 @@
...
+        # Per attention head and per partition values.
+        self.hidden_size_per_partition = projection_size
+        self.hidden_size_per_attention_head = projection_size // config.num_attention_heads
+        self.num_attention_heads_per_partition = config.num_attention_heads
+
+        coeff = None
+        self.norm_factor = math.sqrt(self.hidden_size_per_attention_head)
+        if self.apply_query_key_layer_scaling:
+            coeff = self.layer_number
+            self.norm_factor *= coeff
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:388]: I query_key_layer scaling used in all models ? If this is a different between two models for us we'd rather have that scaling without code pathes

File: src/transformers/models/glm/modeling_glm.py [hunk 17]
```diff
@@ -0,0 +1,1549 @@
...
+        self.coeff = coeff
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:389]: this is not very explicit

File: src/transformers/models/glm/modeling_glm.py [hunk 18]
```diff
@@ -0,0 +1,1549 @@
...
+        # change view [b, np, sq, hn]
+        context_layer = context_layer.reshape(*output_size)
+        # [b, np, sq, hn] --> [b, sq, np, hn]
+        context_layer = context_layer.transpose(1, 2).contiguous()
+        # [b, sq, np, hn] --> [b, sq, hp]
+        new_context_layer_shape = context_layer.size()[:-2] + (self.hidden_size_per_partition,)
+        context_layer = context_layer.reshape(*new_context_layer_shape)
+
+        return context_layer
+
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:462]: Let's standardize and use the GemmaAttention as abaze. 
The only difference seems to be attention softmax in float32, which can be controlled by a flag when calling the softmax instead. Let's not pre allocate and re-use what already exists please

File: src/transformers/models/glm/modeling_glm.py [hunk 19]
```diff
@@ -0,0 +1,1549 @@
...
+    rope_cache = rope_cache.view(-1, 1, sq, xshaped.size(3), 2)
+    x_out2 = torch.stack(
+        [
+            xshaped[..., 0] * rope_cache[..., 0] - xshaped[..., 1] * rope_cache[..., 1],
+            xshaped[..., 1] * rope_cache[..., 0] + xshaped[..., 0] * rope_cache[..., 1],
+        ],
+        -1,
+    )
+    x_out2 = x_out2.flatten(3)
+    return torch.cat((x_out2, x_pass), dim=-1)
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:489]: this is very weird:
1. if you need a variable rotary dimension, it should be set in the config with https://github.com/huggingface/transformers/blob/0fdea8607d7e01eb0e38a1ebeb7feee30a22f0cf/src/transformers/models/persimmon/modeling_persimmon.py#L206 
see this:
https://github.com/huggingface/transformers/blob/0fdea8607d7e01eb0e38a1ebeb7feee30a22f0cf/src/transformers/models/persimmon/modeling_persimmon.py#L309-L318

Comment (ArthurZucker) [reply src/transformers/models/glm/modeling_glm.py:489]: + the cos and sin are not explicit, bbut the computation is the same as what we usually have

File: src/transformers/models/glm/modeling_glm.py [hunk 20]
```diff
@@ -0,0 +1,1549 @@
...
+            query_layer, indices_q, cu_seqlens_q, max_seqlen_in_batch_q = unpad_input(query_layer, attention_mask)
+
+        return (
+            query_layer,
+            key_layer,
+            value_layer,
+            indices_q,
+            (cu_seqlens_q, cu_seqlens_k),
+            (max_seqlen_in_batch_q, max_seqlen_in_batch_k),
+        )
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:581]: smae comment here this was refactored

File: src/transformers/models/glm/modeling_glm.py [hunk 21]
```diff
@@ -0,0 +1,1549 @@
...
+                                                                             dropout_p=self.config.attention_dropout if self.training else 0.0)
+        else:
+            context_layer = torch.nn.functional.scaled_dot_product_attention(query_layer, key_layer, value_layer,
+                                                                             attention_mask,
+                                                                             dropout_p=self.config.attention_dropout if self.training else 0.0)
+        context_layer = context_layer.transpose(1, 2).contiguous()
+        new_context_layer_shape = context_layer.size()[:-2] + (self.hidden_size_per_partition,)
+        context_layer = context_layer.reshape(*new_context_layer_shape)
+        return context_layer
+
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:605]: please let's just go with the usual full class as it is the case for gemma and Llama. yes you re-use the code more here, but it's less explicit and we want explicity in transformers!@

File: src/transformers/models/glm/modeling_glm.py [hunk 22]
```diff
@@ -0,0 +1,1549 @@
...
+                and attention_mask.device.type == "cuda"
+                and not output_attentions
+        ):
+            # Attend to all tokens in fully masked rows in the causal_mask, for example the relevant first rows when
+            # using left padding. This is required by F.scaled_dot_product_attention memory-efficient attention path.
+            # Details: https://github.com/pytorch/pytorch/issues/110213
+            causal_mask = AttentionMaskConverter._unmask_unattended(causal_mask, min_dtype)
+
+        return causal_mask
+
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:736]: why is this here and not in the GlmModel?

File: src/transformers/models/glm/modeling_glm.py [hunk 23]
```diff
@@ -0,0 +1,1549 @@
...
+        self.word_embeddings = nn.Embedding(self.vocab_size, self.hidden_size, dtype=config.torch_dtype, device=device)
+        self.fp32_residual_connection = config.fp32_residual_connection
+
+    def forward(self, input_ids):
+        words_embeddings = self.word_embeddings(input_ids)
+        embeddings = words_embeddings
+        # If the input flag for fp32 residual connection is set, convert for float.
+        if self.fp32_residual_connection:
+            embeddings = embeddings.float()
+        return embeddings
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:755]: no please, let's remove wrapper that are not necessary, this is just a normal embedding

File: src/transformers/models/glm/modeling_glm.py [hunk 24]
```diff
@@ -0,0 +1,1549 @@
...
+
+
+class GLMBlock(torch.nn.Module):
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:758]: Let's standardize, this is a GlmDecoderLayer, let's take inspiration from what already exists

File: src/transformers/models/glm/modeling_glm.py [hunk 25]
```diff
@@ -0,0 +1,1549 @@
...
+        # Second residual connection.
+        if self.apply_residual_connection_post_layernorm:
+            residual = layernorm_output
+        else:
+            residual = layernorm_input
+
+        output = torch.nn.functional.dropout(mlp_output, p=self.hidden_dropout, training=self.training)
+        output = residual + output
+
+        return output, past_key_value
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:826]: there are too many code paths here, for use each codepath is a different model. I don't mind you adding 2 models, but we don't want codepaths like this!

File: src/transformers/models/glm/modeling_glm.py [hunk 26]
```diff
@@ -0,0 +1,1549 @@
...
+
+        self.fp32_residual_connection = config.fp32_residual_connection
+        self.post_layer_norm = config.post_layer_norm
+
+        # Number of layers.
+        self.num_hidden_layers = config.num_hidden_layers
+
+        # Transformer layers.
+        def build_layer(layer_number):
+            return GLMBlock(config, layer_number, device=device)
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:843]: no let's remove one liners

File: src/transformers/models/glm/modeling_glm.py [hunk 27]
```diff
@@ -0,0 +1,1549 @@
...
+        if self.post_layer_norm:
+            LayerNormFunc = GLMRMSNorm if config.rmsnorm else LayerNorm
+            # Final layer norm before output.
+            self.final_layernorm = LayerNormFunc(config.hidden_size, eps=config.layernorm_epsilon, device=device,
+                                                 dtype=config.torch_dtype)
+
+        self.gradient_checkpointing = False
+
+    def _get_layer(self, layer_number):
+        return self.layers[layer_number]
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:856]: no one liner please

File: src/transformers/models/glm/modeling_glm.py [hunk 28]
```diff
@@ -0,0 +1,1549 @@
...
+        self.rotary_pos_emb = GLMRotaryEmbedding(
+            rotary_dim // 2,
+            rope_ratio=config.rope_ratio,
+            original_impl=True,
+            device=device,
+            dtype=config.torch_dtype
+        )
+        self.encoder = init_method(GLMTransformer, config, **init_kwargs)
+        self.output_layer = init_method(nn.Linear, config.hidden_size, config.vocab_size, bias=False,
+                                        dtype=config.torch_dtype, **init_kwargs)
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:1030]: no inits should be done here.

File: src/transformers/models/glm/modeling_glm.py [hunk 29]
```diff
@@ -0,0 +1,1549 @@
...
+
+        batch_size, seq_length = inputs_embeds.shape[:2]
+
+        if use_cache and not isinstance(past_key_values, Cache):  # kept for BC (non `Cache` `past_key_values` inputs)
+            return_legacy_cache = True
+            past_key_values = DynamicCache.from_legacy_cache(past_key_values)
+            logger.warning_once(
+                "We detected that you are passing `past_key_values` as a tuple and this is deprecated and will be removed in v4.43. "
+                "Please use an appropriate `Cache` class (https://huggingface.co/docs/transformers/v4.41.3/en/internal/generation_utils#transformers.Cache)"
+            )
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:1080]: new models should not support legacy behaviour

File: src/transformers/models/glm/modeling_glm.py [hunk 30]
```diff
@@ -0,0 +1,1549 @@
...
+
+        # update position ids
+        if "position_ids" in model_kwargs:
+            position_ids = model_kwargs["position_ids"]
+            new_position_id = position_ids[..., -1:].clone()
+            new_position_id += 1
+            model_kwargs["position_ids"] = torch.cat([position_ids, new_position_id], dim=-1)
+
+        model_kwargs["is_first_forward"] = False
+        return model_kwargs
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:1188]: you should never have to change this form classic LLM

File: src/transformers/models/glm/modeling_glm.py [hunk 31]
```diff
@@ -0,0 +1,1549 @@
...
+        Output shares the same memory storage as `past`.
+        """
+        return tuple(
+            (
+                layer_past[0].index_select(0, beam_idx.to(layer_past[0].device)),
+                layer_past[1].index_select(0, beam_idx.to(layer_past[1].device)),
+            )
+            for layer_past in past
+        )
+
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:1336]: this is also legacy!

File: src/transformers/models/glm/modeling_glm.py [hunk 32]
```diff
@@ -0,0 +1,1549 @@
...
+    Since it does classification on the last token, it requires to know the position of the last token. If a
+    `pad_token_id` is defined in the configuration, it finds the last token that is not a padding token in each row. If
+    no `pad_token_id` is defined, it simply takes the last value in each row of the batch. Since it cannot guess the
+    padding tokens when `inputs_embeds` are passed instead of `input_ids`, it does the same (take the last value in
+    each row of the batch).
+    """,
+    GLM_START_DOCSTRING,
+)
+class GLMForSequenceClassification(GLMPreTrainedModel):
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:1353]: this should be entirely copied from with simple renaming

File: src/transformers/models/glm/modeling_glm.py [hunk 33]
```diff
@@ -0,0 +1,1549 @@
...
+
+
+@add_start_docstrings(
+    """
+    The GLM Model transformer with a token classification head on top (a linear layer on top of the hidden-states
+    output) e.g. for Named-Entity-Recognition (NER) tasks.
+    """,
+    GLM_START_DOCSTRING,
+)
+class GLMForTokenClassification(GLMPreTrainedModel):
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:1469]: same here

File: src/transformers/models/glm/modeling_glm.py [hunk 34]
```diff
@@ -0,0 +1,1549 @@
...
+        )
+
+    def forward(self, hidden_states):
+        # [s, b, 4hp]
+        intermediate_parallel = self.dense_h_to_4h(hidden_states)
+        intermediate_parallel = self.activation_func(intermediate_parallel)
+        # [s, b, h]
+        output = self.dense_4h_to_h(intermediate_parallel)
+        return output
+
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:349]: this is the same as https://github.com/huggingface/transformers/blob/0fdea8607d7e01eb0e38a1ebeb7feee30a22f0cf/src/transformers/models/persimmon/modeling_persimmon.py#L171-L184

**Review by ArthurZucker — COMMENTED** [review]
Hey! Thanks for your work. 
For now my best recommendation is to truly isolate the changes that make this model different from the other 20 "llama-like" (Qwen, Cohere, Persimon, Gemma, Mistral, Phi, etc etc) models we already have in the library? 
- are you using partial rotary embedding? 
- is the order of residual different 

From reviewing the code, it seems like it can be compatible 1-1 with the Persimmon model no?

File: docs/source/en/model_doc/glm.md (hunk 2 above)
Comment (zRzRzRzRzRzRzR) [reply docs/source/en/model_doc/glm.md]: fix now

File: docs/source/en/model_doc/glm.md (hunk 1 above)
Comment (zRzRzRzRzRzRzR) [reply docs/source/en/model_doc/glm.md:42]: finish

File: src/transformers/convert_slow_tokenizer.py (hunk 3 above)
Comment (zRzRzRzRzRzRzR) [reply src/transformers/convert_slow_tokenizer.py]: done

File: src/transformers/models/glm/modeling_glm.py (hunk 34 above)
Comment (zRzRzRzRzRzRzR) [reply src/transformers/models/glm/modeling_glm.py]: `add_bias` will used and I change the code to fit PersimmonMLP

File: src/transformers/models/glm/tokenization_glm.py
Comment (ArthurZucker) [inline src/transformers/models/glm/tokenization_glm.py:1]: this file should be removed as we can map the GPT2Tokenizer direct and use it

File: src/transformers/models/glm/tokenization_glm_fast.py
Comment (ArthurZucker) [reply src/transformers/models/glm/tokenization_glm_fast.py:1]: same comment here, we can use GPT2TokenizerFast!

File: src/transformers/models/glm/configuration_glm.py [hunk 35]
```diff
@@ -0,0 +1,137 @@
...
+"""GLM model configuration"""
+
+from ...configuration_utils import PretrainedConfig
+from ...utils import logging
+
+
+logger = logging.get_logger(__name__)
+
+
+class GLMConfig(PretrainedConfig):
```
Comment (ArthurZucker) [inline src/transformers/models/glm/configuration_glm.py:25]: There is still the issue with the camel casing!

File: src/transformers/models/glm/configuration_glm.py [hunk 36]
```diff
@@ -0,0 +1,137 @@
...
+    r"""
+    This is the configuration class to store the configuration of a [`GLMModel`]. It is used to instantiate a Phi-3
+    model according to the specified arguments, defining the model architecture. Instantiating a configuration with the
```
Comment (ArthurZucker) [inline src/transformers/models/glm/configuration_glm.py:28]:
```suggestion
    This is the configuration class to store the configuration of a [`GLMModel`]. It is used to instantiate a GLM
    model according to the specified arguments, defining the model architecture. Instantiating a configuration with the
```

File: src/transformers/models/glm/modeling_glm.py [hunk 37]
```diff
@@ -0,0 +1,1532 @@
...
+    replace_return_docstrings,
+)
+from .configuration_glm import GLMConfig
+
+
+if is_flash_attn_2_available():
+    from flash_attn.bert_padding import index_first_axis, pad_input, unpad_input  # noqa
+    from flash_attn import flash_attn_func, flash_attn_varlen_func
+
+    _flash_supports_window_size = "window_size" in list(inspect.signature(flash_attn_func).parameters)
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:51]: again this was refactored

File: src/transformers/models/glm/modeling_glm.py [hunk 38]
```diff
@@ -0,0 +1,1532 @@
...
+def _get_unpad_data(attention_mask):
+    seqlens_in_batch = attention_mask.sum(dim=-1, dtype=torch.int32)
+    indices = torch.nonzero(attention_mask.flatten(), as_tuple=False).flatten()
+    max_seqlen_in_batch = seqlens_in_batch.max().item()
+    cu_seqlens = F.pad(torch.cumsum(seqlens_in_batch, dim=0, dtype=torch.int32), (1, 0))
+    return (
+        indices,
+        cu_seqlens,
+        max_seqlen_in_batch,
+    )
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:68]: same comment here

File: src/transformers/models/glm/modeling_glm.py [hunk 39]
```diff
@@ -0,0 +1,1532 @@
...
+        cache = torch.stack([torch.cos(idx_theta), torch.sin(idx_theta)], dim=-1).to(dtype=dtype)
+        return cache
+
+    def forward(self, max_seq_len, offset=0):
+        return self.forward_impl(
+            max_seq_len,
+            self.dim,
+            dtype=self.inv_freq.dtype,
+            device=self.inv_freq.device,
+        )
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:130]: Again same comment here, this is equivalent to LlamaRotaryEmbedidng

File: src/transformers/models/glm/modeling_glm.py (hunk 7 above)
Comment (ArthurZucker) [reply src/transformers/models/glm/modeling_glm.py:161]: This comment is still waiting!

File: src/transformers/models/glm/modeling_glm.py [hunk 40]
```diff
@@ -0,0 +1,1532 @@
...
+        self.multi_query_attention = config.multi_query_attention
+        self.hidden_size = config.hidden_size
+        self.qkv_hidden_size = 3 * self.projection_size
+        self.add_qkv_bias = config.add_qkv_bias
+
+        if self.multi_query_attention:
+            self.num_multi_query_groups_per_partition = self.multi_query_group_num
+            self.qkv_hidden_size = (
+                self.projection_size + 2 * self.hidden_size_per_attention_head * self.multi_query_group_num
+            )
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modeling_glm.py:184]: again same comment about GQA and MQA

**Review by ArthurZucker — COMMENTED** [review]
I am stopping the review as a LOT of the comments are still not adressed

File: src/transformers/models/glm/configuration_glm.py (hunk 35 above)
Comment (zRzRzRzRzRzRzR) [reply src/transformers/models/glm/configuration_glm.py:25]: GLM is the name of our model, not Glm. Do we need to stick to camel case in this context as well?

File: src/transformers/models/glm/configuration_glm.py (hunk 36 above)
Comment (zRzRzRzRzRzRzR) [reply src/transformers/models/glm/configuration_glm.py]: Fix

File: src/transformers/models/glm/modeling_glm.py (hunk 37 above)
Comment (zRzRzRzRzRzRzR) [reply src/transformers/models/glm/modeling_glm.py]: Fix now

File: src/transformers/models/glm/modeling_glm.py (hunk 38 above)
Comment (zRzRzRzRzRzRzR) [reply src/transformers/models/glm/modeling_glm.py:64]: Fix

File: src/transformers/models/glm/configuration_glm.py (hunk 35 above)
Comment (ArthurZucker) [reply src/transformers/models/glm/configuration_glm.py:25]: Yes, it's the same for LLaMa which we set to Llama!

**Review by ArthurZucker — COMMENTED** [review]
🙏🏻

**Comment by ArthurZucker** [comment]
Feel free to ping me again for a review!

**Comment by Cyrilvallez** [comment]
BTW @zRzRzRzRzRzRzR, I took over and am currently adding the model. You can find the new PR here https://github.com/huggingface/transformers/pull/33729, should be ready pretty soon
